### PR TITLE
display NULL instead of empty string

### PR DESCRIPTION
**Pull request by JasonLi-cn**
# Which issue does this PR close?

Closes https://github.com/apache/arrow-rs/issues/2474#issue-1341118778

# Rationale for this change
 

# What changes are included in this PR?

# Are there any user-facing changes?

The user may need to modify the test cases where use array_value_to_string() function, e.g. datafusion

### CODE REVIEW

**Review by tustvold — APPROVED** [review]
I support this change, but I want to give other people the opportunity to weigh in here so will hold off on merging for a bit

**Comment by JasonLi-cn** [comment]
> I support this change, but I want to give other people the opportunity to weigh in here so will hold off on merging for a bit

Ok, does anyone else need to do anything extra here?

**Comment by JasonLi-cn** [comment]
> I support this change, but I want to give other people the opportunity to weigh in here so will hold off on merging for a bit

Ok, thanks @tustvold

**Review by viirya — COMMENTED** [review]
One question. For string value "NULL" mixed with NULL value, does it display the distinction?

File: arrow/src/util/pretty.rs [hunk 1]
```diff
@@ -155,14 +155,14 @@ mod tests {
         let table = pretty_format_batches(&[batch])?.to_string();
 
         let expected = vec![
```
Comment (alamb) [inline arrow/src/util/pretty.rs:157]: this type of change will be needed in much of the DataFusion test base I predict.

**Review by alamb — COMMENTED** [review]
I think this is a reasonable change too but just FYI It will likely result in a substantial diff / effort to upgrade DataFusion and other downstream consumers (like https://github.com/influxdata/influxdb_iox)


I also think we should get some more eyes on this (maybe we can write a note to the mailing list / slack asking people to weigh in)

**Comment by JasonLi-cn** [comment]
> One question. For string value "NULL" mixed with NULL value, does it display the distinction?

I realized that later, so we need to talk about it.

**Comment by alamb** [comment]
Yeah, the problem is that any choice for a string representation of NULL will, by definition, also be a valid string. So the question becomes what string is the least bad to mask / least likely to be in other data.

Maybe `'<NULL>'` ? But really they are all non-ideal ;(

**Comment by tustvold** [comment]
Perhaps we could add an option to specify the NULL string, allowing people to opt-in to the new behaviour.

Similarly we could add an option to quote strings, this would then provide an unambiguous NULL representation.

What do people think?

**Review by Jimexist — APPROVED** [review]
looks good to me but can this be configurable?

**Comment by JasonLi-cn** [comment]
> Perhaps we could add an option to specify the NULL string, allowing people to opt-in to the new behaviour. This would likely involve creating a `FormatOptions` struct and passing it to `pretty_format_batches_with_options` or something
> 
> Similarly we could add an option to quote strings, this would then provide an unambiguous NULL representation.
> 
> What do people think?

I agree :grimacing:

**Comment by alamb** [comment]
> Similarly we could add an option to quote strings, this would then provide an unambiguous NULL representation.

I agree this would be great. 

Maybe we could take the opportunity to make a more extendable API too. What do you all think about something like:

```rust
struct PrettyPrinter {
 // details here
}

impl PrettyPrinter {
  /// Configure which string is used to represent `NULL` (default is "")
  pub fn with_null(mut self, null_string: impl Into<String>) -> Self {
    self.null_string = null_string;
    self
  }

  /// Configure which character, if any, is used to escape strings
  pub fn with_escape(mut self, escape: Option<char>) -> Self {
    self.escape = escape;
    self
  }

   /// Create a visual representation of record batches
   pub fn pretty_format_batches(results: &[RecordBatch]) -> Result<impl Display> {
     // details
   }
  // .. more details here
}
```

Then we could implement the existing API in terms of it like

```rust



///! Create a visual representation of record batches
pub fn pretty_format_batches(results: &[RecordBatch]) -> Result<impl Display> {
  PrettyPrinter::new()
    .with_null(""),
    .with_escape(None)
    .format_batches(results)
}
```


I can file this as a ticket if it seems reasonable

**Comment by tustvold** [comment]
Closed by #3647

**Pull request closed without merging**